zcash_client_backend/proto/proposal.proto

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) 2023 The Zcash developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or https://www.opensource.org/licenses/mit-license.php .
+
+syntax = "proto3";
+package cash.z.wallet.sdk.ffi;
+
+// A data structure that describes the inputs to be consumed and outputs to
+// be produced in a proposed transaction.
+message Proposal {
+    uint32 protoVersion = 1;
+    // ZIP 321 serialized transaction request
+    string transactionRequest = 2;
+    // The unique identifier and amount for each proposed input
+    repeated ProposedInput inputs = 3;
+    // The total value, fee amount, and change outputs of the proposed
+    // transaction
+    TransactionBalance balance = 4;
+    // The fee rule used in constructing this proposal
+    FeeRule feeRule = 6;
+    // The minimum number of confirmations required for transaction inputs to be selected.
+    // Each of the proposed inputs is know to have at least this many confirmations.
```

Comment:
```suggestion
    // Each of the proposed inputs must have at least this many confirmations.
```

Reply:
The draft ZIP 315 [proposes](https://github.com/zcash/zips/pull/607/files#diff-81a3ab3f008443ec864f316692e2e6a15808c87d22f0032a930ce76019a02a05R134-R147) a different number of confirmations for "trusted" and "untrusted" TXOs:
> ## Trusted and untrusted TXOs
> 
> A wallet SHOULD treat received TXOs that are outputs of transactions created
by the same wallet, as trusted TXOs. Wallets MAY enable users to mark specific
external transactions as trusted, allowing their received TXOs also to be
classified as trusted TXOs.
> 
> A wallet SHOULD have a policy that is clearly communicated to the user for
the number of confirmations needed to spend untrusted and trusted TXOs
respectively. The following confirmation policy is RECOMMENDED:
> 
> * 10 confirmations, for untrusted TXOs;
> * 3 confirmations, for trusted TXOs.

So this should be moved to `ProposedInput`, I think.

Reply:
Instead of this being on `ProposedInput` it should probably be an object with `trusted_min_confirmations` and `untrusted_min_confirmations`. Implementing the ZIP 315 recommendation will require changes to the note selection APIs, because at present all of them only support a single `min_confirmations` value and don't reflect the ZIP 315 distinction.

Reply:
Actually I'm wrong; it is sufficient to use a single `minConfirmations` here. There are two reasons why:
1. Draft ZIP 315 is wrong in its conclusion that a higher confirmation depth should be used for untrusted shielded inputs. (["A wallet MUST NOT attempt to spend a TXO in a user-initiated transaction that is not confirmed-spendable."](https://github.com/zcash/zips/pull/607/files#r982891435), where "confirmed-spendable" enforces a higher confirmation depth for untrusted TXOs, including shielded ones.)
2. Even if draft ZIP 315 were right, `minConfirmations` is only a double-check that all of the proposed inputs have at least the given confirmation depth. It wouldn't stop a transaction creation algorithm from enforcing a greater confirmation depth on some inputs if that were useful (which it can be *for fully transparent transactions*).


## Why draft ZIP 315 is wrong

First let's consider a purely transparent, purely UTXO-based blockchain. The term "confirmation depth" is used in two related but distinct ways:

:one: When a wallet reports how much money has been received (either in total or from a given source), counting only TXOs with a given confirmation depth avoids relying on money that the wallet holder might not actually have in the case of a rollback.

:two: When creating a transaction, enforcing a confirmation depth on input selection reduces the risk of the transaction becoming invalid because the chain has rolled back past an input.

There are accidental rollbacks (a.k.a. orphan blocks), but we can reasonably assume that the confirmation depth for *untrusted* TXOs is enough to mitigate transaction unreliability due to those. So the question of whether we need to have separate confirmation depths for trusted vs untrusted inputs depends only on the possibility of rollback *attacks*.

A rollback attack is this: if an adversary has sent coins and obtained benefit from them being understood as received, it has an incentive to roll back the chain in order to also have the coins (instead of the recipient).

This incentive applies whenever a recipient has understood the coins to have been received and provided a benefit. It does not *directly* depend on the confirmation depth at which those inputs are ever used. However, a complicating factor is that if the recipient spent those coins just before the attack, then rolling back the chain far enough for the adversary to benefit will also roll back those spends by the recipient. This may affect the reputation of the recipient with counterparties to whom they spend those coins, and they will likely want to avoid that problem. As I understand it, that is the motivation for using the same confirmation depth for :one: and :two: in a transparent cryptocurrency (and may be a reason why :one: and :two: tend to be conflated in discussions).

----

Now consider how this changes in the case of a Zerocash-style cryptocurrency that uses anchors to identify commitment treestates. :one: is the same, and :two: still applies to transparent inputs. However, for transactions that are at-least-partly shielded, we *also* have:

:three: When creating an at-least-partly shielded transaction, enforcing a minimum depth on anchor selection reduces the risk of the transaction becoming invalid because the chain has rolled back past the anchor.

Is it ever beneficial to enforce a higher confirmation depth when choosing shielded inputs than the anchor depth [*]? No, because this will not change whether or not the transaction succeeds: that depends only on anchor invalidation (under the assumption that the probability of getting the same commitment treestate after a rollback is negligible [**]). Indeed, the consensus cannot tell how far back the shielded notes were output. And so the argument linking :one: and :two: above does not work to link :one: and :three:.

> [*] [In v5 transactions there is at most one Sapling anchor and at most one Orchard anchor. In v4 transactions there are separate anchors for each Sapling input.](https://zips.z.cash/protocol/protocol.pdf#txnencoding) By "anchor depth" I mean the minimum depth of any anchor in the transaction. In any case, we never want to choose different anchors for different Sapling Spends, because that leaks information :— potentially a lot of information if the reason we are doing it is to allow including an input that could not otherwise be included. Similarly, there is never any reason to choose Sapling and Orchard anchors at different depths in a mixed-pool transaction.

> [**] We can't actually infer that it is negligible because an attack might only affect fully transparent transactions, leaving all updates to commitment trees the same. However, that doesn't really affect the argument because we can't *rely* on the commitment treestates remaining the same.

Note that in transparent chains, when a transaction is in one of the rolled-back blocks in a rollback attack, it still has a good chance of getting into the new chain quickly (if not already included by the attacker), as long as its inputs are still present in the new chain. This doesn't apply to at-least-partly shielded transactions, which will have to be recreated if any anchor is invalidated.

And so, it's sufficient to have a single `minConfirmations` field. (Actually, `minTargetHeight` would be sufficient on its own, if we trust the transaction creator to calculate it correctly to enforce all of the confirmation depths. On the other hand, explicitly specifying `minConfirmations` is arguably less prone to error.)

We should also change the ZIP 315 draft to reflect the above reasoning (using the concept of "trusted" and "untrusted" TXOs for balance reporting but not for creation of at-least-partially shielded transactions).

Reply:
The above argument has been reproduced and refined at https://github.com/zcash/zips/pull/607#issuecomment-1707243952 ; please read that version since there are subtle differences (in particular, there is no also benefit in choosing a greater confirmation depth for *transparent* inputs of a partly-shielded transaction than the anchor depth).